en/_posts/2024-05-23-understanding-github-artifact-attestations.md

```diff
@@ -0,0 +1,224 @@
+doesn’t meet this requirement for [SLSA Build
+L3](https://slsa.dev/spec/v1.0/levels#build-l3-hardened-builds):
+
+prevent secret material used to sign the provenance from being accessible to the user-defined build steps.
+
+Because GitHub Actions are run in the same job VM with other build steps there
+is a chance that the user-defined build steps could access the secret material,
+namely the certificate’s private key provided by Fulcio, and use it for
+nefarious purposes. Normally an attacker, by compromising the build, might be
+able to use this key material to make up their own provenance for a malicious
+artifact and sign it with the key.
+
+However, as we’ll see, this kind of attack is mostly mitigated by using Sigstore.
+
+## SLSA Build L2+?
+
+By using Sigstore’s Fulcio the certificate used to sign the provenance contains
+much of the provenance information in its OID claims. These claims are signed
+by part of Fulcio instance’s certificate chain and thus cannot be modified by
+the user-defined build steps unless the Sigstore instance or GitHub’s OIDC
+provider are compromised.
+
+So while the SLSA provenance itself might be modified the certificate OID
+claims cannot. So GitHub can check the OID claims against the expected values
+in order to verify them even though the user-defined build steps had access to
+the signing key. So this is why verification doesn’t rely on the provenance
+itself and instead relies on the certificate’s OID claims for verification.
+
+Some folks have colloquially referred to this combination of Sigstore and SLSA
+Build L2 as SLSA Build L2+ since it provides some of the benefits of SLSA Build
+L3 without actually fulfilling all of the requirements of L3.
+
+## Limits of SLSA Build L2+
+
+However, this comes with a caveat. This means that if the user-defined build
+steps could have access to the signing keys then **only** the information in
+the certificate’s OID claims are trustworthy. No information that isn’t
+included in these claims can be included in the provenance because it can’t be
+verified later.
+
+Crucially this includes values like the GitHub Actions workflow
+[`inputs`](https://docs.github.com/en/actions/learn-github-actions/contexts#inputs-context)
+and
+[`vars`](https://docs.github.com/en/actions/learn-github-actions/contexts#vars-context)
+contexts. These include the inputs into a build and could potentially be used
+for script injection attacks. While they can’t be used to attack the provenance
+itself, they could be used to attack the build without leaving any trace in the
+provenance.
+
+This means that while GitHub Artifact Attestations can be used with policy
+engines/tools like Rego, Cue, and OPA, these policies can’t be evaluated on the
+build inputs. For example, to verify that the build had no inputs etc. in order
+to prevent these kinds of attacks.
+
+This architecture also relies heavily on Sigstore’s Fulcio Certificate
```

Comment:
it'd be fun to show the equivalent cosign-based flow to do the same, from a security PoV. Basically, install cosign via their Action, then sign with cosign your predicate.

Reply:
Yeah, it's basically the same as our SLSA container generator. I'll add those commands as examples.